Modules/_sqlite/connection.c

```diff
@@ -911,32 +953,22 @@ static int _authorizer_callback(void* user_arg, int action, const char* arg1, co
     PyGILState_STATE gilstate = PyGILState_Ensure();
 
     PyObject *ret;
-    int rc;
+    int rc = SQLITE_DENY;
 
-    ret = PyObject_CallFunction((PyObject*)user_arg, "issss", action, arg1, arg2, dbname, access_attempt_source);
+    callback_context *ctx = (callback_context *)user_arg;
+    assert(ctx != NULL);
+    ret = PyObject_CallFunction(ctx->callable, "issss", action, arg1, arg2,
+                                dbname, access_attempt_source);
 
     if (ret == NULL) {
-        pysqlite_state *state = pysqlite_get_state(NULL);
-        if (state->enable_callback_tracebacks) {
-            PyErr_Print();
-        }
-        else {
-            PyErr_Clear();
-        }
-
+        print_or_clear_traceback(ctx);
```

Comment:
I just stumbled over a bug in these three callbacks:
They all check `enable_callback_tracebacks` and either `PyErr_Print` (and clear) the current exception, or `PyErr_Clear` the current exception. That's fine for the function, step, and final callbacks; they all pass an error back to SQLite, and we translate that into an exception after checking the return value of `sqlite3_step()` or `sqlite3_finalize()`. However, for the callbacks we're dealing with in this PR, we _always_ clear the current exception:

```
Python 3.10.0rc1 (v3.10.0rc1:cc115e5bf6, Aug  2 2021, 17:54:14) [Clang 12.0.5 (clang-1205.0.22.11)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> cx = sqlite3.connect(":memory:")
>>> cx.set_trace_callback(lambda x: 5/0)
>>> try:
...     cx.execute("select 1")
... except:
...     print("exception occurred")
... else:
...     print("no exception occurred")
... 
<sqlite3.Cursor object at 0x1006dc640>
no exception occurred
>>> sqlite3.enable_callback_tracebacks(True)
>>> try: 
...     cx.execute("select 1")
... except:
...     print("exception occurred")
... else:
...     print("no exception occurred")
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <lambda>
ZeroDivisionError: division by zero
<sqlite3.Cursor object at 0x100891c40>
no exception occurred
```

I cannot imagine that this was the intended behaviour.

Reply:
I opened #28133 for this.